docker.go

```diff
@@ -371,15 +376,46 @@ func checkImage(image string) bool {
 	return out.String() != ""
 }
 
+// ECR Regex: https://regex101.com/r/GRxU06/1
+var reECR = regexp.MustCompile(`(?P<account>[0-9]+)\.dkr\.ecr\.(?P<region>[a-z0-9\-]+)\.amazonaws\.com`)
+
 func refreshImage(image string) {
 	ErrPrintf("Checking if there is a newer version of docker image %v\n", image)
-	dockerUpdateCmd := exec.Command("docker", "pull", image)
-	dockerUpdateCmd.Stdout, dockerUpdateCmd.Stderr = os.Stderr, os.Stderr
-	must(dockerUpdateCmd.Run())
+	err := getDockerUpdateCmd(image).Run()
+	if err != nil {
+		matches, _ := utils.MultiMatch(image, reECR)
+		account, accountOk := matches["account"]
+		region, regionOk := matches["region"]
+		if accountOk && regionOk && awsConfigExist() {
+			ErrPrintf("Failed to pull %v. It is an ECR image, trying again after a login.\n", image)
+			loginToECR(account, region)
+			must(getDockerUpdateCmd(image).Run())
+		} else {
+			panic(err)
+		}
+	}
 	touchImageRefresh(image)
 	ErrPrintln()
 }
 
+func loginToECR(account string, region string) {
+	awsSession := session.Must(session.NewSessionWithOptions(session.Options{SharedConfigState: session.SharedConfigEnable}))
+	svc := ecr.New(awsSession, &aws.Config{Region: aws.String(region)})
+	requestInput := &ecr.GetAuthorizationTokenInput{RegistryIds: []*string{aws.String(account)}}
+	result := must(svc.GetAuthorizationToken(requestInput)).(*ecr.GetAuthorizationTokenOutput)
+
+	decodedLogin := string(must(base64.StdEncoding.DecodeString(*result.AuthorizationData[0].AuthorizationToken)).([]byte))
+	dockerUpdateCmd := exec.Command("docker", "login", "-u", strings.Split(decodedLogin, ":")[0],
```

Comment:
Here, it would be nice to use the docker library instead of calling directly docker.

I try to get rid as much as possible of `exec.Command("docker", ...`.

Reply:
No, I did not try. It was just a suggestion.

Then I prefer your solution. I suppose that the docker exec also store some information somewhere to make the login persistent. We may look at it later.

Reply:
I looked around and it seems like what people do is pass the auth directly to the `ImagePull` func. I don't know if the login is cached at that time, I'll merge this and try to replace it later.